improver/developer_tools/metadata_interpreter.py

```diff
@@ -0,0 +1,578 @@
+        verbose (bool):
+            Optional flag to include information about the source of the
+            metadata interpretation (eg name, coordinates, attributes, etc)        
+
+    Returns:
+        str:
+            Formatted string describing metadata in human-readable form
+    """
+    if interpreter.unhandled:
+        return f"{interpreter.diagnostic} is not handled by this interpreter"
+
+    def vstring(source_metadata):
+        """Format additional message for verbose output"""
+        return f"    Source: {source_metadata}"
+
+    def format_standard_cases(interpreter):
+        """Format prob / perc / diagnostic information"""
+        field_type = interpreter.field_type.replace("_", " ")
+        diagnostic = interpreter.diagnostic.replace("_", " ")
+        if interpreter.relative_to_threshold:
+            relative_to_threshold = interpreter.relative_to_threshold.replace("_", " ")
+
+        rval = []
+        rtt = (
+            f" {relative_to_threshold} thresholds"
+            if interpreter.field_type == interpreter.PROB
+            else ""
+        )
+        rval.append(f"It contains {field_type} of {diagnostic}{rtt}")
+        if verbose:
+            rval.append(vstring("name, threshold coordinate (probabilities only)"))
+
+        if interpreter.methods:
+            rval.append(f"These {field_type} are of {diagnostic}{interpreter.methods}")
+            if verbose:
+                rval.append(vstring("cell methods"))
+
+        ppstring = "some" if interpreter.post_processed else "no"
+        rval.append(f"It has undergone {ppstring} significant post-processing")
+        if verbose:
+            rval.append(vstring("title attribute"))
+        return rval
+
+    field_type = interpreter.field_type.replace("_", " ")
+    output = []
+    output.append(f"This is a {interpreter.prod_type} {field_type} file")
+    if verbose:
+        output.append(vstring("name, coordinates"))
+
+    if interpreter.diagnostic not in SPECIAL_CASES:
+        output.extend(format_standard_cases(interpreter))
+
+    if interpreter.diagnostic in ANCILLARIES:
+        output.append(f"This is a static ancillary with no time information")
+    elif interpreter.blended:
+        output.append(f"It contains blended data from models: {interpreter.model}")
+        if verbose:
+            output.append(vstring("title attribute, model ID attribute"))
+    else:
+        if interpreter.model:
+            output.append(f"It contains data from {interpreter.model}")
+            if verbose:
+                output.append(vstring("model ID attribute"))
+        else:
+            output.append("It has no source model information and cannot be blended")
+            if verbose:
+                output.append(vstring("model ID attribute (missing)"))
+
+    if interpreter.warnings:
+        warning_string = "\n".join(interpreter.warnings)
+        output.append(f"WARNINGS:\n{warning_string}")
+
+    return "\n".join(output)
```

Comment:
I think it'd be neater here to return a list of strings from this function rather than one long string containing line breaks. That would also avoid the need for expected output strings in tests to contain `\n` line breaks.

The CLI wrapper can handle line breaks, rather than this function doing most but not all of it. A final line break is currently (correctly) added by the `print(output)` in line 67 of improver/developer_tools/run_metadata_interpreter.py

Reply:
I mostly wrote this function to handle the formatting aspect, which to my mind includes the output being clearly spaced and readable.  A user then has a choice of whether to run this on a file from the wrapper, or to script it and print the output from `display_interpretation` directly.  Is there a functional reason why it's better not to do it this way?

Reply:
Currently, the output of this function is mostly-but-not-quite formatted text, so it's not a good interface between components.

Background: On UNIX-like systems (and the POSIX standards) text consists of zero or more lines. Each line consists of zero or more characters, followed by a line break (eg. `\n`). This definition of text allows for handy features such as easy parsing and composability of text. For example, the `cat` command can join together multiple text files to make one large text file by just joining the bytes together rather than needing to do any special handling at the end of files.

Two options I can think of are:
- Change this function to return a list of strings (the existing `output` variable). Change the CLI function to loop through the list of strings and print each string. Python's print function defaults add a line break at the end of each line.
- Change this function to add a newline character at the end of the last line. Change the CLI function to print with no additional line break eg. `print(out, end="")`

Reply:
I've moved the script to a piece of follow-on work following Stephen's comments, however I'm happy to add a newline to the end of the output from the display function.